src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/ReadsKey.java

```diff
@@ -22,33 +25,124 @@ public static int hashKeyForFragment(int strandedUnclippedStart, boolean reverse
 
-        key = 31 * key + ReadUtils.getReferenceIndex(second, header);
-        key = 31 * key + ReadUtils.getStrandedUnclippedStart(second);
-        return 31 * key + (second.isReverseStrand() ? 0 : 1);
+    public static ReadsKey getKeyForPair(final SAMFileHeader header, final GATKRead first, final GATKRead second, final Map<String, Byte> libraryKeyMap) {
+        return new keyForPair(longKeyForFragment(ReadUtils.getStrandedUnclippedStart(first),
+                                first.isReverseStrand(),
+                                ReadUtils.getReferenceIndex(first, header),
+                                libraryKeyMap.get(ReadUtils.getLibrary(first, header, LibraryIdGenerator.UNKNOWN_LIBRARY))),
+                longKeyForPair(ReadUtils.getStrandedUnclippedStart(second),
+                        second.isReverseStrand(),
+                        ReadUtils.getReferenceIndex(second, header))
+                );
     }
 
     /**
-     * Makes a unique key for the read.
+     *
      */
-    public static String keyForRead(final GATKRead read) {
-        return read.getName();
+    public static class keyForFragment extends ReadsKey {
+        final long keyValue;
+
+        keyForFragment(final long key) {
+            this.keyValue = key;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            keyForFragment that = (keyForFragment) o;
+            return keyValue == that.keyValue;
+        }
+
+        @Override
+        public int hashCode() {
+
+            return Objects.hash(keyValue);
+        }
+
+        @Override
+        public String toString() {
+            return Long.toString(keyValue);
+        }
     }
 
     /**
-     * Makes a hash key for the read.
+     *
      */
-    public static int hashKeyForRead(final GATKRead read) {
-        return read.getName().hashCode();
+    public static class keyForPair extends ReadsKey {
+        final long fragmentValue;
+        final long keyValue;
+
+        keyForPair(final long fragmentValue, final long pairValue) {
+            this.fragmentValue = fragmentValue;
+            this.keyValue = pairValue;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            keyForPair that = (keyForPair) o;
+            return fragmentValue == that.fragmentValue &&
+                    keyValue == that.keyValue;
+        }
+
+        @Override
+        public int hashCode() {
+
+            return Objects.hash(fragmentValue, keyValue);
+        }
+
+        @Override
+        public String toString() {
+            return fragmentValue + " " + keyValue;
+        }
+    }
+
+
+    private static long longKeyForFragment(int strandedUnclippedStart, boolean reverseStrand, int referenceIndex, byte library) {
```

Comment:
It still seems weird to encode these as longs when we could either A) keep them as the individual values and let serialization handle it, or B) convert them to a byte array and save a few bits.  We can revisit later if we think it would help I guess.  Metrics seem higher priority than shaving a few more bits here.

Reply:
Do we have a list written somewhere?